lib/ec2/managers/EbsManager.rb

```diff
@@ -0,0 +1,176 @@
+require "common/manager/Manager"
+require "conf/Configuration"
+require "ec2/EC2"
+require "ec2/loaders/EbsLoader"
+require "ec2/models/EbsGroupConfig"
+require "ec2/models/EbsGroupDiff"
+require "util/StatusCodes"
+
+require "aws-sdk"
+
+module Cumulus
+  module EC2
+    class EbsManager < Common::Manager
+
+      def initialize
+        super()
+        @create_asset = true
+        @client = Aws::EC2::Client.new(region: Configuration.instance.region, profile: Configuration.instance.profile)
+      end
+
+      def resource_name
+        "EBS Volume Group"
+      end
+
+      def local_resources
+        @local_resources ||= Hash[EbsLoader.groups.map { |local| [local.name, local] }]
+      end
+
+      def aws_resources
+        @aws_resources ||= EC2::group_ebs_volumes
+      end
+
+      def unmanaged_diff(aws)
+        EbsGroupDiff.unmanaged(aws)
+      end
+
+      def added_diff(local)
+        EbsGroupDiff.added(local)
+      end
+
+      def diff_resource(local, aws)
+        local.diff(aws)
+      end
+
+      EbsMigrationData = Struct.new(:cumulus_group, :set_group_vols)
+      def migrate
+        puts Colors.blue("Migrating EBS Volume Groups...")
+
+        # Create the directories
+        ec2_dir = "#{@migration_root}/ec2"
+        ebs_dir = "#{ec2_dir}/ebs"
+
+        if !Dir.exists?(@migration_root)
+          Dir.mkdir(@migration_root)
+        end
+        if !Dir.exists?(ec2_dir)
+          Dir.mkdir(ec2_dir)
+        end
+        if !Dir.exists?(ebs_dir)
+          Dir.mkdir(ebs_dir)
+        end
+
+        puts "Would you like Cumulus to automatically update your volumes to have a Group tag that matches the name of the instance they are attached to? (y/n)"
+        update_tags = (STDIN.getc.downcase[0] == "y")
+
+        # Migrate any volumes that have already been grouped
+        vol_groups = Hash[EC2.group_ebs_volumes.map { |group_name, cumulus_group| [group_name, EbsMigrationData.new(cumulus_group, nil)] }]
+
+        # Use the instance name to group volumes if they do not have a group.  Anything
+        # not attached should not get migrated
+        migratable_vols = EC2.ebs_volumes.select { |vol| vol.group.nil? and !vol.attachments.empty? and vol.attached? }
+        instance_grouped = migratable_vols.group_by do |vol|
+          attachments = vol.attachments.select { |att| att.state == "attached" || att.state == "attaching" }
+          attachments.first.instance_id
```

Comment:
the line above with `migratable_vols` ensures there is at least one attachment with the correct state.  It's a little obscure that the check is being done I know, and this is why doing things like select+map at the same time would be so nice